org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/InlayHintVisitor.java

```diff
@@ -0,0 +1,261 @@
+				hints.add(hint);
+			}
+		} catch (JavaModelException e) {
+			JavaLanguageServerPlugin.logException(e.getMessage(), e);
+		}
+	}
+
+	/**
+	 * Get parameter names from the method binding.
+	 * @param methodBinding
+	 * @return the parameter names, or <code>null</code> if it's not available.
+	 */
+	private String[] getParameterNames(IMethodBinding methodBinding) {
+		if (!hasSource(methodBinding)) {
+			return null;
+		}
+		String[] parameterNames = null;
+		IMethod method = (IMethod) methodBinding.getJavaElement();
+		if (method != null) {
+			try {
+				parameterNames = method.getParameterNames();
+			} catch (JavaModelException e) {
+				JavaLanguageServerPlugin.logException(e);
+			}
+		} else if (methodBinding.getDeclaringClass().isRecord()) {
+			// get the record constructor parameter names via its declared fields
+			IVariableBinding[] declaredFields = methodBinding.getDeclaringClass().getDeclaredFields();
+			parameterNames = Arrays.stream(declaredFields).map(field -> {
+				return field.getName();
+			}).toArray(String[]::new);
+		}
+		return parameterNames;
+	}
+
+	/**
+	 * Check if the given method has attached source.
+	 */
+	private boolean hasSource(IMethodBinding methodBinding) {
+		IType type = (IType) methodBinding.getDeclaringClass().getJavaElement();
+		if (type == null) {
+			return false;
+		}
+
+		ITypeRoot typeRoot = type.getTypeRoot();
+		if (typeRoot == null || !typeRoot.exists()) {
+			return false;
+		}
+
+		try {
+			return typeRoot.getBuffer() != null;
+		} catch (JavaModelException e) {
+			JavaLanguageServerPlugin.logException(e.getMessage(), e);
+		}
+		return false;
+	}
+
+	/**
+	 * Check if the argument needs to show an inlay hint for it. The following cases will not
+	 * have inlay hints:
+	 * <p>
+	 * <ul>
+	 *   <li>It's in literal mode but the given argument is not a literal</li>
+	 *   <li>Argument name and parameter names are same</li>
+	 * </ul>
+	 * </p>
+	 */
+	private boolean acceptArgument(Expression argument, String paramName) {
+		if (InlayHintsParameterMode.LITERALS.equals(preferenceManager.getPreferences().getInlayHintsParameterMode())) {
+			if (!isLiteral(argument)) {
+				return false;
+			}
+		}
+
+		return !isSameName(argument, paramName);
+	}
+
+	/**
+	 * Check whether the argument is literal type.
+	 */
+	private boolean isLiteral(Expression argument) {
+		if (argument instanceof BooleanLiteral || argument instanceof CharacterLiteral ||
+				argument instanceof NullLiteral || argument instanceof NumberLiteral ||
+				argument instanceof StringLiteral || argument instanceof TypeLiteral) {
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Check whether the argument name is the same as the parameter name.
+	 * Note: if the argument is a cast expression, only the expression
+	 * to be casted will be compared.
+	 */
+	private boolean isSameName(Expression argument, String paramName) {
```

Comment:
@rgrunber For this part, I suggest we can first only do an exact match when deciding to hide the inlay hint. And according to users' feedback, we can consider adding a setting to control the behavior in the future, something like `exact match`, `contained`, `etc...`

Reply:
That's fine. Actually, one of the disadvantages for suppressing the hints for substring matches is users may not immediately understand why some parameter hints are hidden, and may see it as a bug.